be/src/vec/functions/function_string.h

```diff
@@ -2601,43 +2591,62 @@ class FunctionUrlDecode : public IFunction {
-                        size_t input_rows_count) const override {
+    Status execute_impl(FunctionContext* context, Block& block, const ColumnNumbers& arguments,
+                        size_t result, size_t input_rows_count) const override {
         auto res = ColumnString::create();
-        auto& res_offsets = res->get_offsets();
-        auto& res_chars = res->get_chars();
-        res_offsets.resize(input_rows_count);
+        res->get_offsets().reserve(input_rows_count);
 
-        ColumnPtr argument_column =
-                block.get_by_position(arguments[0]).column->convert_to_full_column_if_const();
-        const auto* url_col = check_and_get_column<ColumnString>(argument_column.get());
+        const auto* url_col =
+                assert_cast<const ColumnString*>(block.get_by_position(arguments[0]).column.get());
 
-        if (!url_col) {
-            return Status::InternalError("Not supported input argument type");
+        std::string decoded_url;
+        for (size_t i = 0; i < input_rows_count; ++i) {
+            auto url = url_col->get_data_at(i);
+            if (!url_decode(url.to_string(), &decoded_url)) {
+                return Status::InternalError("Decode url failed");
+            }
+            res->insert_data(decoded_url.data(), decoded_url.size());
+            decoded_url.clear();
         }
 
-        std::string decoded_url;
+        block.get_by_position(result).column = std::move(res);
+        return Status::OK();
+    }
+};
 
-        for (size_t i = 0; i < input_rows_count; ++i) {
-            auto source = url_col->get_data_at(i);
-            StringRef url_val(const_cast<char*>(source.data), source.size);
+class FunctionUrlEncode : public IFunction {
+public:
+    static constexpr auto name = "url_encode";
+    static FunctionPtr create() { return std::make_shared<FunctionUrlEncode>(); }
+    String get_name() const override { return name; }
+    size_t get_number_of_arguments() const override { return 1; }
+    DataTypePtr get_return_type_impl(const DataTypes& arguments) const override {
+        return std::make_shared<DataTypeString>();
+    }
 
-            url_decode(url_val.to_string(), &decoded_url);
+    Status execute_impl(FunctionContext* context, Block& block, const ColumnNumbers& arguments,
+                        size_t result, size_t input_rows_count) const override {
+        auto res = ColumnString::create();
+        res->get_offsets().reserve(input_rows_count);
 
-            StringOP::push_value_string(decoded_url, i, res_chars, res_offsets);
-            decoded_url.clear();
+        const auto* url_col =
+                assert_cast<const ColumnString*>(block.get_by_position(arguments[0]).column.get());
+
+        std::string encoded_url;
+        for (size_t i = 0; i < input_rows_count; ++i) {
+            auto url = url_col->get_data_at(i);
+            if (!url_encode(url.to_string_view(), &encoded_url)) {
+                return Status::InternalError("Encode url failed");
```

Comment:
when will this happen? if for invalid inputs, it should be `InvalidArgument`

Reply:
Because `url_encode` may return error, so handle it just in case something wrong.

Reply:
We use `curl_easy_escape` to impl url_encode, `curl_easy_escape` will return null when some error happened.